### PR TITLE
add basic index test

### PR DESCRIPTION
**Pull request by LHT129**
- use basic index test to test all kind of index (hnsw, diskann, ...)
- add some classic dataset to reuse for different test
- mark some index static and store status for different exception
      section tests

### CODE REVIEW

File: tests/fixtures/fixtures.cpp [hunk 1]
```diff
@@ -45,19 +47,26 @@ normalize(float* input_vector, int64_t dim) {
...
+        32,   33,  48,   // sse(32) + generic(dim < 16)
+        64,   65,  70,   // avx(64) + generic(dim < 16)
+        96,   97,  109,  // avx(64) + sse(32) + generic(dim < 16)
+        128,  129,       // avx512(128) + generic(dim < 16)
+        160,  161,       // avx512(128) + sse(32) + generic(dim < 16)
+        192,  193,       // avx512(128) + avx(64) + generic(dim < 16)
+        224,  225,       // avx512(128) + avx(64) + sse(32) + generic(dim < 16)
+        256,  512,       // common used dims
+        784,  960,       // common used dims
+        1024, 1536};     // common used dims
```
Comment (wxyucs) [inline tests/fixtures/fixtures.cpp]: add 2048 into common used dims

File: tests/fixtures/fixtures.cpp [hunk 2]
```diff
@@ -267,4 +276,30 @@ GenTestItems(uint64_t count, uint64_t max_length, uint64_t max_index) {
...
+}
+
+vsag::DatasetPtr
+generate_one_dataset(int64_t dim, uint64_t count) {
+    auto result = vsag::Dataset::Make();
+    auto [ids, vectors] = generate_ids_and_vectors(count, dim, true, time(nullptr));
+    result->Dim(dim)
+        ->NumElements(count)
+        ->Float32Vectors(CopyVector(vectors))
+        ->Ids(CopyVector(ids))
```
Comment (wxyucs) [inline tests/fixtures/fixtures.cpp:294]: why we need copy vectors and ids here

File: tests/fixtures/test_dataset.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+#include "vsag/dataset.h"
+namespace fixtures {
+
+class TestDataset {
+public:
+    using DatasetPtr = vsag::DatasetPtr;
+
+    TestDataset(DatasetPtr base, DatasetPtr query, DatasetPtr ground_truth)
+        : base_(base), query_(query), ground_truth_(ground_truth){};
```
Comment (wxyucs) [inline tests/fixtures/test_dataset.h]: use const reference or std::move

File: tests/test_diskann.cpp [hunk 4]
```diff
@@ -12,17 +12,50 @@
...
 #include <nlohmann/json.hpp>
 
+#include "test_index.h"
 #include "vsag/errors.h"
 #include "vsag/vsag.h"
 
 const std::string tmp_dir = "/tmp/";
 
+TEST_CASE_METHOD(fixtures::TestIndex, "diskann build test", "[ft][index][diskann]") {
+    vsag::Options::Instance().logger()->SetLevel(vsag::Logger::Level::kDEBUG);
```
Comment (wxyucs) [inline tests/test_diskann.cpp:30]: is there some way to set the logger with debug mode in the first line of each test case

File: tests/fixtures/fixtures.cpp (hunk 2 above)
Comment (LHT129) [reply tests/fixtures/fixtures.cpp:294]: owner = true, make the lifetime manage by dataset

File: tests/fixtures/fixtures.cpp (hunk 1 above)
Comment (LHT129) [reply tests/fixtures/fixtures.cpp]: done

File: tests/fixtures/test_dataset.h (hunk 3 above)
Comment (LHT129) [reply tests/fixtures/test_dataset.h]: done

File: tests/test_diskann.cpp (hunk 4 above)
Comment (LHT129) [reply tests/test_diskann.cpp:30]: add in constructor of fixtures::TestIndex

**Review by wxyucs — APPROVED** [review]
lgtm

File: tests/fixtures/fixtures.cpp [hunk 5]
```diff
@@ -267,4 +276,30 @@ GenTestItems(uint64_t count, uint64_t max_length, uint64_t max_index) {
     return result;
 }
 
+template <typename T>
+static T*
+CopyVector(const std::vector<T>& vec) {
```
Comment (jiaweizone) [inline tests/fixtures/fixtures.cpp:281]: minor: CopyVector -> ToArray ?

File: tests/test_hnsw_new.cpp [hunk 6]
```diff
@@ -0,0 +1,1177 @@
...
+#include <nlohmann/json.hpp>
+#include <numeric>
+#include <random>
+
+#include "simd/simd.h"
+#include "test_index.h"
+#include "vsag/errors.h"
+#include "vsag/vsag.h"
+
+TEST_CASE_METHOD(fixtures::TestIndex, "HNSW Build & ContinueAdd Test", "[ft][hnsw]") {
```
Comment (jiaweizone) [inline tests/test_hnsw_new.cpp:32]: `test_hsnw_new.cpp` is used for new VSAG index framework test suite ?

File: tests/fixtures/fixtures.cpp (hunk 5 above)
Comment (LHT129) [reply tests/fixtures/fixtures.cpp:281]: Copy means alloc another memory, but ToArray may loss this meaning

File: tests/test_hnsw_new.cpp (hunk 6 above)
Comment (LHT129) [reply tests/test_hnsw_new.cpp:32]: the test_hnsw_new will replace test_hnsw soon, it's from the basic index test, but only test features what hnsw need

**Review by jiaweizone — APPROVED** [review]
LGTM

**Pull request merged**